<a href="https://colab.research.google.com/github/Olga-Lyt/DTA-2026/blob/main/HW/pandas_tasks_beginners_DTA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практичні завдання з pandas для початківців

**Мета:** Набути базових навичок роботи з бібліотекою pandas для аналізу табличних даних.

**Інструкції:**
- Виконуйте завдання послідовно — кожне наступне спирається на попередні.
- Після виконання кожного завдання напишіть короткий висновок у комірці `**Висновок:**`.
- Не змінюйте код у комірках з підготовкою даних.

**Набір даних:** Умовна таблиця продажів інтернет-магазину за 2023 рік.

---
## Підготовка: Генерація тестових даних

Запустіть цю комірку перед початком роботи. Не змінюйте код.

In [2]:
import pandas as pd
import numpy as np

np.random.seed(42)
n = 200

categories = ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт']
cities = ['Київ', 'Харків', 'Одеса', 'Дніпро', 'Львів']
payment_methods = ['Картка', 'Готівка', 'Онлайн']

df = pd.DataFrame({
    'order_id':        range(1001, 1001 + n),
    'date':            pd.date_range('2023-01-01', periods=n, freq='D')[:n],
    'category':        np.random.choice(categories, n),
    'product':         ['Товар_' + str(i) for i in np.random.randint(1, 51, n)],
    'city':            np.random.choice(cities, n),
    'quantity':        np.random.randint(1, 10, n),
    'price':           np.round(np.random.uniform(50, 5000, n), 2),
    'payment_method':  np.random.choice(payment_methods, n),
    'rating':          np.random.choice([1, 2, 3, 4, 5, np.nan], n, p=[0.05, 0.1, 0.15, 0.3, 0.3, 0.1]),
    'is_returned':     np.random.choice([True, False], n, p=[0.1, 0.9]),
})

# Додаємо кілька пропущених значень
df.loc[np.random.choice(df.index, 10, replace=False), 'city'] = np.nan
df.loc[np.random.choice(df.index, 5, replace=False), 'price'] = np.nan

# Розраховуємо загальну суму
df['total'] = df['quantity'] * df['price']

print('Дані успішно згенеровано. Рядків:', len(df))
print('Колонки:', list(df.columns))

Дані успішно згенеровано. Рядків: 200
Колонки: ['order_id', 'date', 'category', 'product', 'city', 'quantity', 'price', 'payment_method', 'rating', 'is_returned', 'total']


---
## Завдання 1 (Легке). Перший погляд на дані

Виведіть перші 10 рядків датафрейму `df` та отримайте загальну інформацію про нього (кількість рядків, колонок, типи даних) за допомогою методів `head()` та `info()`.

In [3]:
display(df.head(10))
df.info()

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
0,1001,2023-01-01,Продукти,Товар_34,Київ,1,963.38,Картка,5.0,False,963.38
1,1002,2023-01-02,Спорт,Товар_6,Львів,6,1086.28,Картка,4.0,False,6517.68
2,1003,2023-01-03,Книги,Товар_22,Дніпро,4,1883.84,Готівка,5.0,False,7535.36
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52
5,1006,2023-01-06,Одяг,Товар_16,Львів,9,1876.12,Онлайн,NaN,False,16885.08
6,1007,2023-01-07,Книги,Товар_33,Дніпро,4,2339.55,Картка,4.0,False,9358.20
7,1008,2023-01-08,Книги,Товар_9,Львів,4,3749.98,Готівка,4.0,False,14999.92
8,1009,2023-01-09,Книги,Товар_6,Дніпро,6,231.58,Онлайн,4.0,False,1389.48
9,1010,2023-01-10,Спорт,Товар_16,Одеса,3,1299.56,Готівка,1.0,False,3898.68


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        200 non-null    int64         
 1   date            200 non-null    datetime64[ns]
 2   category        200 non-null    object        
 3   product         200 non-null    object        
 4   city            190 non-null    object        
 5   quantity        200 non-null    int64         
 6   price           195 non-null    float64       
 7   payment_method  200 non-null    object        
 8   rating          177 non-null    float64       
 9   is_returned     200 non-null    bool          
 10  total           195 non-null    float64       
dtypes: bool(1), datetime64[ns](1), float64(3), int64(2), object(4)
memory usage: 15.9+ KB


**Висновок:** *(Напишіть, скільки рядків і колонок у датафреймі, які типи даних зустрічаються, чи є пропущені значення)*  
У датафреймі 200 рядків та 11 колонок. Типи даних: bool, datetime64[ns], float64, int64, object. Пропущених значень немає.

---
## Завдання 2 (Легке). Описова статистика

Використайте метод `describe()` для отримання описової статистики числових колонок. Окремо виведіть мінімум, максимум і середнє значення колонки `price`.

In [4]:
display(df.describe())
print(f"Minimum price: {df['price'].min()}")
print(f"Maximum price: {df['price'].max()}")
print(f"Average price: {df['price'].mean().round(2)}")

,order_id,date,quantity,price,rating,total
count,200.000000,200,200.000000,195.000000,177.000000,195.000000
mean,1100.500000,2023-04-10 12:00:00,4.700000,2500.579641,3.813559,11629.097590
min,1001.000000,2023-01-01 00:00:00,1.000000,139.650000,1.000000,149.350000
25%,1050.750000,2023-02-19 18:00:00,2.000000,1288.215000,3.000000,3893.955000
50%,1100.500000,2023-04-10 12:00:00,5.000000,2517.290000,4.000000,8746.840000
75%,1150.250000,2023-05-30 06:00:00,7.000000,3706.975000,5.000000,16922.420000
max,1200.000000,2023-07-19 00:00:00,9.000000,4950.300000,5.000000,44324.640000
std,57.879185,NaN,2.618066,1380.066748,1.169808,9968.293697


Minimum price: 139.65
Maximum price: 4950.3
Average price: 2500.58


**Висновок:** *(Опишіть діапазон цін, середнє значення, чи є суттєвий розкид)*  
Діапазон цін становить від 139,65 до 4950,30. Середня ціна 2500,58. Розкид цін є суттєвим.

---
## Завдання 3 (Легке). Пропущені значення

Знайдіть кількість пропущених значень у кожній колонці за допомогою `isnull().sum()`. Визначте, який відсоток рядків має пропущені значення хоча б в одній колонці (`dropna()` та `len()`).

In [5]:
display(df.isnull().sum())

mis = df.isnull().any(axis=1).sum()
total_rows = len(df)
mis_pct = (mis / total_rows) * 100

print(f"\nКількість рядків з пропущеними значеннями хоча б в одній колонці: {mis}")
print(f"Відсоток рядків з пропущеними значеннями: {mis_pct:.2f}%")

,0
order_id,0
date,0
category,0
product,0
city,10
quantity,0
price,5
payment_method,0
rating,23
is_returned,0



Кількість рядків з пропущеними значеннями хоча б в одній колонці: 37
Відсоток рядків з пропущеними значеннями: 18.50%


**Висновок:** *(У яких колонках є пропуски, наскільки це критично для аналізу)*  
Пропуски є в колонках city та rating. Відсоток рядків з пропущеними значеннями (18,5%) є значущим.

---
## Завдання 4 (Легке). Відбір колонок і рядків

Створіть новий датафрейм `df_subset`, що містить лише колонки `order_id`, `category`, `city`, `total`. Потім виведіть рядки з індексами від 10 до 20 включно за допомогою `.iloc[]`.

In [6]:
# Ваш код тут
df_subset = df[['order_id', 'category', 'city', 'total']]
df_subset.iloc[10:21, 1:]
df_subset[['order_id', 'category', 'city', 'total']]
df_subset.iloc[10:21]


,order_id,category,city,total
10,1011,Продукти,Дніпро,21486.48
11,1012,Книги,Львів,31368.89
12,1013,Спорт,Харків,7748.40
13,1014,Одяг,Дніпро,18787.72
14,1015,Продукти,NaN,1741.50
15,1016,Одяг,Одеса,4529.38
16,1017,Продукти,Київ,10745.84
17,1018,Спорт,Одеса,10001.84
18,1019,Електроніка,Дніпро,12444.75
19,1020,Продукти,Харків,13422.92


**Висновок:** *(Поясніть різницю між вибором колонок через список і через `iloc`)*  
При виборі колонок через список застосовуються назви колонок, а при виборі через iloc їх індекси.

---
## Завдання 5 (Легке). Фільтрація даних

Відфільтруйте датафрейм і отримайте:
1. Всі замовлення з категорії `'Електроніка'`.
2. Всі замовлення, де `price` перевищує 2000.
3. Поверненi замовлення (`is_returned == True`) з міста `'Київ'`.

Виведіть кількість рядків для кожного фільтру.

In [7]:
# Ваш код тут
orders_electronics = df[df['category'] == 'Електроніка']
print(f"Кількість замовлень з категорії 'Електроніка': {len(orders_electronics)}")

orders_price_over_2000 = df[df['price'] > 2000]
print(f"Кількість замовлень, де price > 2000: {len(orders_price_over_2000)}")

returned_orders_kyiv = df[(df['is_returned'] == True) & (df['city'] == 'Київ')]
print(f"Кількість повернутих замовлень з міста 'Київ': {len(returned_orders_kyiv)}")

Кількість замовлень з категорії 'Електроніка': 43
Кількість замовлень, де price > 2000: 116
Кількість повернутих замовлень з міста 'Київ': 7


**Висновок:** *(Яка частка замовлень відповідає кожному критерію? Що можна сказати про повернення з Києва?)*  
Частка замовлень з категорії 'Електроніка': 43 / 200 * 100 = 21,5%  
Частка замовлень, де price > 2000: 116 / 200 *100 = 58%  
Частка повернутих замовдень з Києва до всіх замовлень: 7 / 200 * 100 = 3,5%  
З 200 замовлень 7 були повернені з Києва, що становить 3.5% від усіх замовлень.





---
## Завдання 6 (Середнє). Сортування та ранжування

1. Відсортуйте датафрейм за колонкою `total` у спадному порядку та виведіть топ-10 найбільших замовлень.
2. Додайте нову колонку `rank`, що містить порядковий номер замовлення за розміром суми (від найбільшого до найменшого) за допомогою `.rank()`.

In [8]:
# Ваш код тут

df['rank'] = df['total'].rank(ascending=False, method = 'min')
df.sort_values('total', ascending=False).head(10)

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total,rank
190,1191,2023-07-10,Одяг,Товар_22,Київ,9,4924.96,Готівка,4.0,True,44324.64,1.0
58,1059,2023-02-28,Продукти,Товар_33,Харків,9,4848.88,Картка,5.0,False,43639.92,2.0
186,1187,2023-07-06,Спорт,Товар_34,Київ,9,4675.12,Готівка,5.0,False,42076.08,3.0
55,1056,2023-02-25,Електроніка,Товар_3,Дніпро,8,4912.77,Картка,5.0,False,39302.16,4.0
169,1170,2023-06-19,Продукти,Товар_50,Київ,8,4739.05,Готівка,5.0,False,37912.40,5.0
191,1192,2023-07-11,Спорт,Товар_25,Харків,9,4202.55,Готівка,4.0,True,37822.95,6.0
193,1194,2023-07-13,Одяг,Товар_22,NaN,8,4608.17,Картка,4.0,False,36865.36,7.0
51,1052,2023-02-21,Продукти,Товар_35,Дніпро,8,4488.99,Онлайн,4.0,False,35911.92,8.0
25,1026,2023-01-26,Книги,Товар_18,Дніпро,8,4458.11,Картка,3.0,False,35664.88,9.0
143,1144,2023-05-24,Продукти,Товар_31,Київ,9,3732.53,Готівка,5.0,False,33592.77,10.0


In [9]:
max_total = df['total'].max()
print(f"Максимальна сума замовлення: {max_total:.2f}")

top_10_categories = df.sort_values('total', ascending=False).head(10)['category'].unique()
print(f"Категорії найбільших замовлень (топ-10): {', '.join(top_10_categories)}")

Максимальна сума замовлення: 44324.64
Категорії найбільших замовлень (топ-10): Одяг, Продукти, Спорт, Електроніка, Книги


**Висновок:** *(Яка максимальна сума замовлення? З яких категорій найбільші замовлення?)*  
Максимальна сума замовлення: 44324.64
Категорії найбільших замовлень (топ-10): Одяг, Продукти, Спорт, Електроніка, Книги

---
## Завдання 7 (Середнє). Групування та агрегація

За допомогою `groupby()` розрахуйте для кожної категорії товарів:
- загальний обсяг продажів (сума `total`),
- кількість замовлень,
- середню ціну (`price`).

Відсортуйте результат за загальним обсягом продажів.

In [10]:
# Ваш код тут
category_stats = df.groupby('category').agg(
    total_sales=('total', 'sum'),
    order_count=('order_id', 'count'),
    average_price=('price', 'mean')
).sort_values(by='total_sales', ascending=False)

display(category_stats)

,total_sales,order_count,average_price
category,,,
Продукти,583761.68,49,2513.979787
Одяг,474135.61,35,2707.906176
Електроніка,428129.79,43,2334.470698
Спорт,400802.56,36,2494.205143
Книги,380844.39,37,2491.881944


**Висновок:** *(Яка категорія приносить найбільший дохід? Чи збігається це з кількістю замовлень?)*  
Найбільший дохід приносить категорія Продукти, що збігається з кількістю замовлень.

---
## Завдання 8 (Середнє). Обробка пропущених значень

1. Заповніть пропущені значення в колонці `price` медіаною цін відповідної категорії (використайте `groupby().transform('median')`).
2. Заповніть пропущені значення в колонці `city` рядком `'Невідомо'`.
3. Пропущені значення в `rating` залиште без змін — поясніть чому у висновку.

Перевірте, що пропусків у `price` та `city` більше немає.

In [11]:
# Ваш код тут
df['price'] = df['price'].fillna(df.groupby('category')['price'].transform('median'))
df['city'] = df['city'].fillna('Невідомо')

df['total'] = df['quantity'] * df['price']

df['rank'] = df['total'].rank(ascending=False, method = 'min')

print("Перевірка пропущених значень після обробки:")
display(df.isnull().sum())

Перевірка пропущених значень після обробки:


,0
order_id,0
date,0
category,0
product,0
city,0
quantity,0
price,0
payment_method,0
rating,23
is_returned,0


**Висновок:** *(Чому медіана краща за середнє для заповнення цін? Чому пропуски в rating можна залишити?)*  
Медіана краща за середнє при заповненні цін, тому що вона не враховує викиди, а середнє розраховується з урахуванням всіх значень, у т.ч. і викидів. Пропуски в rating можна залишити, тому що клієнт може  не мати бажання оцінювати товар.

---
## Завдання 9 (Середнє). Додавання обчислюваних колонок

Додайте до датафрейму нові колонки:
1. `month` — місяць замовлення (з колонки `date`).
2. `price_category` — категорія ціни: `'Низька'` (до 500), `'Середня'` (500–2000), `'Висока'` (понад 2000). Використайте `pd.cut()` або `np.select()`.
3. `discount_total` — сума зі знижкою 15%.

In [12]:
# Ваш код тут
df['month'] = df['date'].dt.month

bins = [0, 500, 2000, np.inf]
labels = ['Низька', 'Середня', 'Висока']
df['price_category'] = pd.cut(df['price'], bins=bins, labels=labels, right=False)

df['discount_total'] = df['total'] * 0.85

display(df.head())

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total,rank,month,price_category,discount_total
0,1001,2023-01-01,Продукти,Товар_34,Київ,1,963.38,Картка,5.0,False,963.38,193.0,1,Середня,818.8730
1,1002,2023-01-02,Спорт,Товар_6,Львів,6,1086.28,Картка,4.0,False,6517.68,121.0,1,Середня,5540.0280
2,1003,2023-01-03,Книги,Товар_22,Дніпро,4,1883.84,Готівка,5.0,False,7535.36,113.0,1,Середня,6405.0560
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39,176.0,1,Висока,2081.1315
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52,29.0,1,Висока,18506.6420


**Висновок:** *(Який розподіл цінових категорій? Як метод pd.cut спрощує категоризацію?)*  
Ціни розподіляються за наступними категоріями: низька (до 500), середня (500-2000), висока (понад 2000). Метод pd.cut дозволяє без складних операторів визначити межі діапазонів, що робить код більш читабельним та зменшує вірогідність помилок.

---
## Завдання 10 (Середнє). Зведена таблиця (pivot table)

Побудуйте зведену таблицю за допомогою `pd.pivot_table()`, де:
- рядки — `city`,
- колонки — `category`,
- значення — сума `total`,
- агрегація — `sum`.

Додайте підсумки по рядках і колонках (`margins=True`).

In [13]:
# Ваш код тут
pivot_table = pd.pivot_table(df,
                             index='city',
                             columns='category',
                             values='total',
                             aggfunc='sum',
                             margins=True)

display(pivot_table)

category,Електроніка,Книги,Одяг,Продукти,Спорт,All
city,,,,,,
Дніпро,65438.16,116118.970,31179.870,147336.56,72244.38,432317.940
Київ,58713.88,87339.030,103754.430,118283.10,113742.95,481833.390
Львів,154285.60,126188.885,74340.690,55768.80,43155.08,453739.055
Невідомо,21012.44,25984.450,46962.310,40508.03,5917.59,140384.820
Одеса,84611.65,26773.570,90218.245,60466.75,65539.34,327609.555
Харків,44068.06,15669.180,135537.230,199766.70,105100.00,500141.170
All,428129.79,398074.085,481992.775,622129.94,405699.34,2336025.930


**Висновок:** *(Яке місто і яка категорія дають найбільший сумарний дохід? Що показують margins?)*
Найбільший сумарний дохід дає м. Харків. За категорією найбільший сумарний дохід припадає на продукти. Margins показують  підсумок за кожним рядком та стовпчиком таблиці.

---
## Завдання 11 (Середнє). Підрахунок унікальних значень і частот

1. Підрахуйте кількість унікальних товарів (`product`) у кожній категорії.
2. Виведіть топ-5 найпопулярніших товарів за кількістю замовлень за допомогою `value_counts()`.
3. Визначте частку кожного способу оплати у відсотках (`normalize=True`).

In [14]:
print("Кількість унікальних товарів у кожній категорії:")
display(df.groupby('category')['product'].nunique())

print("\nТоп-5 найпопулярніших товарів за кількістю замовлень:")
display(df['product'].value_counts().head(5))

print("\nЧастка кожного способу оплати у відсотках:")
display(df['payment_method'].value_counts(normalize=True) * 100)

Кількість унікальних товарів у кожній категорії:


,product
category,
Електроніка,32
Книги,25
Одяг,27
Продукти,30
Спорт,25



Топ-5 найпопулярніших товарів за кількістю замовлень:


,count
product,
Товар_33,11
Товар_48,9
Товар_35,9
Товар_22,7
Товар_20,6



Частка кожного способу оплати у відсотках:


,proportion
payment_method,
Готівка,34.0
Онлайн,33.5
Картка,32.5


**Висновок:** *(Який спосіб оплати найпопулярніший? Чи рівномірний розподіл між способами?)*  
Найпопулярніший спосіб оплати готівка (формально). Насправді розподіл між способами оплати рівномірний.

---
## Завдання 12 (Середнє). Аналіз повернень

Розрахуйте рівень повернень (частку повернених замовлень) для кожної категорії та кожного міста. Визначте:
- категорію з найвищим рівнем повернень,
- місто з найнижчим рівнем повернень.

Підказка: використайте `groupby()['is_returned'].mean()`.

In [15]:
# Ваш код тут


In [16]:
return_rate_by_category = df.groupby('category')['is_returned'].mean().sort_values(ascending=False)
print("Рівень повернень за категоріями:")
display(return_rate_by_category)

return_rate_by_city = df.groupby('city')['is_returned'].mean().sort_values(ascending=True)
print("\nРівень повернень за містами:")
display(return_rate_by_city)

category_highest_return = return_rate_by_category.index[0]
city_lowest_return = return_rate_by_city.index[0]

print(f"\nКатегорія з найвищим рівнем повернень: {category_highest_return} ({return_rate_by_category.iloc[0]:.2%})")
print(f"Місто з найнижчим рівнем повернень: {city_lowest_return} ({return_rate_by_city.iloc[0]:.2%})")

Рівень повернень за категоріями:


,is_returned
category,
Одяг,0.171429
Електроніка,0.162791
Книги,0.135135
Спорт,0.111111
Продукти,0.061224



Рівень повернень за містами:


,is_returned
city,
Дніпро,0.052632
Харків,0.076923
Невідомо,0.100000
Львів,0.108696
Київ,0.189189
Одеса,0.233333



Категорія з найвищим рівнем повернень: Одяг (17.14%)
Місто з найнижчим рівнем повернень: Дніпро (5.26%)


**Висновок:** *(Яка категорія найчастіше повертається? Про що це може свідчити?)*  
Найчастіше повертаються товари з категорії Одяг, що може свідчити про складнощі з підбором розміру одягу при замовленні.


---
## Завдання 13 (Підвищена складність). Часовий аналіз

1. Встановіть колонку `date` як індекс датафрейму.
2. Розрахуйте щомісячний обсяг продажів (ресемплінг `resample('ME')` або `resample('M')`, sum по `total`).
3. Знайдіть найкращий і найгірший місяці за продажами.
4. Поверніть звичайний числовий індекс (`reset_index()`).

In [18]:
# Ваш код тут

df_time = df.set_index('date')

monthly_sales = df_time['total'].resample('ME').sum()

display(monthly_sales)

best_month = monthly_sales.idxmax()
worst_month = monthly_sales.idxmin()

print(f"\nМісяць з найбільшими продажами: {best_month.strftime('%Y-%m')} (Total: {monthly_sales.max():.2f})")
print(f"Місяць з найменшими продажами: {worst_month.strftime('%Y-%m')} (Total: {monthly_sales.min():.2f})")

df_time = df_time.reset_index()

,total
date,
2023-01-31,351659.990
2023-02-28,354807.150
2023-03-31,305283.785
2023-04-30,322315.225
2023-05-31,335619.620
2023-06-30,386464.610
2023-07-31,279875.550



Місяць з найбільшими продажами: 2023-06 (Total: 386464.61)
Місяць з найменшими продажами: 2023-07 (Total: 279875.55)


**Висновок:** *(Який місяць виявився найуспішнішим? Чи є сезонність у даних?)*  
Найуспішнішим місяцем виявився червень. Сезонності у даних немає, найменші продажі були у липні.

---
## Завдання 14 (Підвищена складність). Накопичувальні показники

1. Відсортуйте датафрейм за датою.
2. Додайте колонку `cumulative_total` — накопичувальний обсяг продажів з початку року (`cumsum()`).
3. Визначте, в який день було досягнуто 50% від загального річного обсягу продажів.

In [26]:
# Ваш код тут

df_sorted = df.sort_values('date').copy()

df_sorted['cumulative_total'] = df_sorted['total'].cumsum()

total_annual_sales = df_sorted['total'].sum()
half_annual_sales = total_annual_sales * 0.5

day_50_percent_reached = df_sorted[df_sorted['cumulative_total'] >= half_annual_sales]['date'].iloc[0]

display(df_sorted.head())
print(f"\nЗагальний річний обсяг продажів: {total_annual_sales:.2f}")
print(f"50% від річного обсягу продажів ({half_annual_sales:.2f}) було досягнуто: {day_50_percent_reached.strftime('%Y-%m-%d')}")

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total,rank,month,price_category,discount_total,cumulative_total
0,1001,2023-01-01,Продукти,Товар_34,Київ,1,963.38,Картка,5.0,False,963.38,193.0,1,Середня,818.8730,963.38
1,1002,2023-01-02,Спорт,Товар_6,Львів,6,1086.28,Картка,4.0,False,6517.68,121.0,1,Середня,5540.0280,7481.06
2,1003,2023-01-03,Книги,Товар_22,Дніпро,4,1883.84,Готівка,5.0,False,7535.36,113.0,1,Середня,6405.0560,15016.42
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39,176.0,1,Висока,2081.1315,17464.81
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52,29.0,1,Висока,18506.6420,39237.33



Загальний річний обсяг продажів: 2336025.93
50% від річного обсягу продажів (1168012.97) було досягнуто: 2023-04-14


**Висновок:** *(Коли приблизно була зароблена половина річного доходу? Що це означає для бізнесу?)*  
Половина річного доходу була досягнута 14 квітня 2023 року. Це означає, що бізнес розвивається добре, з перевищенням запланованих показників.

---
## Завдання 15 (Підвищена складність). apply() та lambda

1. Використайте `apply()` з lambda-функцією для створення колонки `total_with_tax`, де ПДВ становить 20%, але лише для замовлень на суму понад 1000 (решта — без змін).
2. Напишіть функцію `classify_rating(r)`, що повертає `'Поганий'` (1–2), `'Нейтральний'` (3), `'Хороший'` (4–5), `'Без оцінки'` (NaN). Застосуйте її до колонки `rating`.

In [28]:
# Ваш код тут

df['total_with_tax'] = df['total'].apply(lambda x: x * 1.20 if x > 1000 else x)

def classify_rating(r):
    if pd.isna(r):
        return 'Без оцінки'
    elif r >= 4:
        return 'Хороший'
    elif r >= 3:
        return 'Нейтральний'
    else:
        return 'Поганий'

df['rating_category'] = df['rating'].apply(classify_rating)

display(df[['total', 'total_with_tax', 'rating', 'rating_category']].head())

good_orders = (df['rating_category'] == 'Хороший').sum()
total_orders = len(df)

share_good = good_orders / total_orders * 100

print(f"Кількість замовлень з хорошим рейтингом: {good_orders}")
print(f"Частка замовлень з хорошим рейтингом: {share_good:.2f}%")

,total,total_with_tax,rating,rating_category
0,963.38,963.380,5.0,Хороший
1,6517.68,7821.216,4.0,Хороший
2,7535.36,9042.432,5.0,Хороший
3,2448.39,2938.068,5.0,Хороший
4,21772.52,26127.024,3.0,Нейтральний


Кількість замовлень з хорошим рейтингом: 116
Частка замовлень з хорошим рейтингом: 58.00%


**Висновок:** *(Яка частка замовлень має хороший рейтинг? Коли apply кращий за vectorized-операції?)*  
Кількість замовлень з хорошим рейтингом 58%. Аpply кращий за vectorized-операції у випадках, коли потрібні складні розрахунки. В інших випадках доцільніше використовувати vectorized-операції, тому що вони працюють швидше та будуть більш ефективними.


---
## Завдання 16 (Підвищена складність). Злиття датафреймів (merge)

Створіть додатковий датафрейм `df_discounts` з колонками `category` та `discount_pct` (різні знижки для кожної категорії). Об'єднайте його з основним датафреймом за допомогою `pd.merge()` та розрахуйте `final_total` — суму після застосування відповідної знижки.

```python
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})
```

In [32]:
# Ваш код тут

df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})

df['total'] = df['quantity'] * df['price']

df_merged = pd.merge(
    df,
    df_discounts,
    on='category',
    how='left'
)

df_merged['final_total'] = (
    df_merged['total'] * (1 - df_merged['discount_pct'])
).round(2)

df_merged['discount_amount'] = (
    df_merged['total'] - df_merged['final_total']
).round(2)

discount_by_category = (
    df_merged
    .groupby('category')['discount_amount']
    .sum()
    .sort_values(ascending=False)
)

top_category = discount_by_category.idxmax()
top_discount = discount_by_category.max()

print("Сумарна знижка по категоріях:")
print(discount_by_category)

print(
    f"\nНайбільшу сумарну знижку отримала категорія "
    f"'{top_category}' — {top_discount:.2f}"
)

display(
    df_merged[
        [
            'order_id',
            'category',
            'total',
            'discount_pct',
            'discount_amount',
            'final_total'
        ]
    ].head()
)


Сумарна знижка по категоріях:
category
Одяг           96398.55
Спорт          60854.90
Електроніка    42813.03
Книги          19903.68
Продукти           0.00
Name: discount_amount, dtype: float64

Найбільшу сумарну знижку отримала категорія 'Одяг' — 96398.55


,order_id,category,total,discount_pct,discount_amount,final_total
0,1001,Продукти,963.38,0.00,0.00,963.38
1,1002,Спорт,6517.68,0.15,977.65,5540.03
2,1003,Книги,7535.36,0.05,376.77,7158.59
3,1004,Спорт,2448.39,0.15,367.26,2081.13
4,1005,Спорт,21772.52,0.15,3265.88,18506.64


**Висновок:** *(Яка категорія отримала найбільшу сумарну знижку? Коли варто використовувати merge замість map?)*  
Найбільшу сумарну знижку отримала категорія Одяг. Мerge замість map варто використовувати у випадку, коли потрібно об'єднати декілька таблиць.

---
## Завдання 17 (Підвищена складність). Групові трансформації

1. Додайте колонку `city_avg_total` — середня сума замовлення по місту (для кожного рядка підставте середнє по його місту), використайте `groupby().transform('mean')`.
2. Додайте колонку `above_city_avg` — булева колонка, що показує, чи перевищує замовлення середнє по місту.
3. Підрахуйте частку таких замовлень для кожного міста.

In [37]:
# Ваш код тут
df_merged['city_avg_total'] = (
    df_merged
    .groupby('city')['total']
    .transform('mean')
    .round(2)
)

df_merged['above_city_avg'] = (
    df_merged['total'] > df_merged['city_avg_total']
)

city_share = (
    df_merged
    .groupby('city')['above_city_avg']
    .mean()
    .mul(100)
    .round(2)
    .sort_values(ascending=False)
)

print("Частка замовлень вище середнього по місту (%):")
print(city_share)

display(
    df_merged[
        [
            'order_id',
            'city',
            'total',
            'city_avg_total',
            'above_city_avg'
        ]
    ].head()
)

Частка замовлень вище середнього по місту (%):
city
Харків      43.59
Одеса       43.33
Дніпро      42.11
Невідомо    40.00
Київ        35.14
Львів       34.78
Name: above_city_avg, dtype: float64


,order_id,city,total,city_avg_total,above_city_avg
0,1001,Київ,963.38,13022.52,False
1,1002,Львів,6517.68,9863.89,False
2,1003,Дніпро,7535.36,11376.79,False
3,1004,Дніпро,2448.39,11376.79,False
4,1005,Дніпро,21772.52,11376.79,True


**Висновок:** *(Чим transform відрізняється від звичайного groupby? У якому місті найбільша частка великих замовлень?)*  
Тransform доцільно використовувати тоді, коли потрібно створити нову колонку і порахувати в ній значення. Найбільша частка великих замовлень у Харкові.

---
## Завдання 18 (Складне). Ковзне середнє та аномалії

1. Згрупуйте дані за місяцем та розрахуйте щоденну виручку.
2. Розрахуйте 7-денне ковзне середнє виручки (`rolling(7).mean()`).
3. Визначте дні-аномалії: дні, коли виручка відхилилася від ковзного середнього більш ніж на 2 стандартних відхилення.

In [38]:
# Ваш код тут
daily_revenue = (
    df_merged
    .groupby('date')['total']
    .sum()
    .reset_index()
    .sort_values('date')
)

daily_revenue['month'] = daily_revenue['date'].dt.to_period('M')

daily_revenue['rolling_mean_7d'] = (
    daily_revenue['total']
    .rolling(window=7)
    .mean()
)

daily_revenue['rolling_std_7d'] = (
    daily_revenue['total']
    .rolling(window=7)
    .std()
)

daily_revenue['upper_bound'] = (
    daily_revenue['rolling_mean_7d']
    + 2 * daily_revenue['rolling_std_7d']
)

daily_revenue['lower_bound'] = (
    daily_revenue['rolling_mean_7d']
    - 2 * daily_revenue['rolling_std_7d']
)

daily_revenue['is_anomaly'] = (
    (daily_revenue['total'] > daily_revenue['upper_bound']) |
    (daily_revenue['total'] < daily_revenue['lower_bound'])
)

anomalies = daily_revenue[daily_revenue['is_anomaly']]

print(f"Кількість аномальних днів: {len(anomalies)}")

display(
    anomalies[
        [
            'date',
            'total',
            'rolling_mean_7d',
            'upper_bound',
            'lower_bound'
        ]
    ]
)

Кількість аномальних днів: 3


,date,total,rolling_mean_7d,upper_bound,lower_bound
51,2023-02-21,35911.92,10910.347143,33602.890485,-11782.196199
85,2023-03-27,29701.80,8022.478571,29291.938774,-13246.981631
143,2023-05-24,33592.77,9505.597143,31257.099478,-12245.905193


**Висновок:** *(Скільки аномальних днів виявлено? Що може бути причиною різких відхилень виручки?)*  
Виявлено три аномальних днів. Причиною різких відхилень виручки може бути безліч факторів, наприклад: знижки, свята, рекламні кампанії, оптові покупці та ін.

---
## Завдання 19 (Складне). Когортний аналіз

Виконайте спрощений когортний аналіз за місяцем першого замовлення:
1. Для кожного товару (`product`) знайдіть місяць першого продажу — це його «когорта».
2. Об'єднайте цю інформацію з основним датафреймом.
3. Побудуйте зведену таблицю: когорта (місяць першого продажу) проти місяця замовлення, значення — кількість замовлень.

Підказка: використайте `groupby('product')['date'].transform('min')` та `dt.to_period('M')`.

In [24]:
# Ваш код тут


**Висновок:** *(Що показує когортна таблиця? Чи продовжують товари з першої когорти продаватися в наступних місяцях?)*

---
## Завдання 20 (Складне). Підсумковий звіт

Сформуйте підсумковий звіт у вигляді одного датафрейму `summary_report`, що містить для кожної категорії:
- загальний дохід (`total_revenue`),
- кількість замовлень (`order_count`),
- середній чек (`avg_order_value`),
- частку доходу від загального (`revenue_share_pct`),
- рівень повернень у відсотках (`return_rate_pct`),
- середній рейтинг без NaN (`avg_rating`).

Відсортуйте за `total_revenue` спадно. Збережіть результат у файл `summary_report.csv`.

In [25]:
# Ваш код тут


**Висновок:** *(Яка категорія є лідером за доходом? Чи є зв'язок між рейтингом та рівнем повернень? Які рекомендації можна дати бізнесу на основі цього звіту?)*

---
## Загальні підсумки

Після виконання всіх завдань дайте відповідь на такі питання:

1. Які методи pandas виявилися для вас найбільш корисними?
2. Які труднощі виникли під час роботи з пропущеними значеннями?
3. Що можна додати до аналізу, щоб зробити його більш повним?

*(Ваші відповіді тут)*